### ISSUE TITLE
Classloading problem with WebApplication classloader in OSGi environment

### ISSUE THREAD

**Issue by pieterjanpintens** [eclipse/jetty.project#4419]
**Jetty version**
9.4.20 (eclipse 4.13)
**Java version**
Java 8 AWS corretto 
**Question**
We use Jetty in our application to serve web applications.
When upgrading we noticed that we are now seeing class loading problems from our application code in our webapplications. We tracked it down to changes related to the Webapplication classloader. That is the method where this classloader decides if the class loaded is a server class or not. We think that the flow is correct but that some utility methods might be off.

In detail.

The classloading starts in:
org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(String, boolean)
With isParentLoaderPriority = false.

In the section under `// Try the parent loader`
Our parent classloader can load the class but the _context.isServerClass(parentClass)
in the underlying if return true so the class is not returned and a class not found exception is thrown.

The class we are loading in the OSGi Bundle class so it is definitely not a server class (it wont match with the package patters specified by jetty). The reason why this fails lies deeper. To decide if a class is a server class, at a certain point, the classloading code starts looking at the location (the URI where the class is loaded from). Here we hit problems.

**If we install our app (put our jars) in a folder with a space we hit the problem if not we don't hit the problem.**

This is related to the following two methods:

**org.eclipse.jetty.webapp.ClasspathPattern.combine(IncludeExcludeSet<Entry, String>, String, IncludeExcludeSet<Entry, URI>, Supplier<URI>)**
This method is called to determine if a class is a server class. It calls `URI uri = location.get();`
This call will result in the following method being called:

**org.eclipse.jetty.util.TypeUtil.getCodeSourceLocation(Class<?>)**
This method must return the URI of the source location (jar file/class file).
The problem here is that calling `toURI` on a URL with spaces will throw URISyntaxException.
As a result null is returned (even if the URL points to a valid file:// location).
This explains why we only hit the problem when we have spaces in our paths of our jars

Because the latter returns null and locations is empty the return value of the combine (and eventually the check for our class being a server class) will become true. So the WebApplication classloader will decided not to return our class even if it is definitely not a server class.

So the question:

Is this a bug on your end or should we try to work around it by mangling with the ClassPatterns we can set our our web application context? To me it feels that the utility classes are not doing a good job which results in poor behavior of the webapp class loader.

1. The TypeUtil.getCodeSourceLocation can do better, the url is pointing to a correct file, it should be possible to make this a proper URI without throwing a URISyntaxException.
2. One can argue that the ClasspathPattern.combine should not return true if URI is null and locations is empty. This is whitelist vs blacklist.

Fixing either of these will solve the problem.

**Comment by janbartel** [comment]
Can you try with 9.4.24? I recall we fixed a bug in this area recently. I'm away from my computer right now, but will post a link to the issue tomorrow.

**Comment by joakime** [comment]
> If we install our app (put our jars) in a folder with a space we hit the problem if not we don't hit the problem.

See #2575 - you likely have one of the OSGi implementations that has the URL/URI escaping bug in `Bundle.getEntry()`

**Comment by pieterjanpintens** [comment]
I don't think this is the problem, the class is loaded fine, it is the jetty web application classloader that decided to throw away the class because it assumes it calculates incorrectly that it is a server class. Our setup is probably not that common since we have a custom parent classloader to delegate class loading in our OSGi environment.

The fix suggested in the issue above to convert a URL with scheme file:// to a URI using the File class can also be applied to the code in this [method](https://github.com/eclipse/jetty.project/blob/fe1a3c67a1f67d59b7e768d351c95cce52d6fb36/jetty-util/src/main/java/org/eclipse/jetty/util/TypeUtil.java#L706). i think this will resolve the problem. 

In other words using this to convert from URL to URI.
```
private static URI toFileURI(URL fileUrl) {
	if (!fileUrl.getProtocol().equalsIgnoreCase("file")) {
		throw new IllegalArgumentException("not file: scheme");
	} else {
		return new File(fileUrl.getFile()).toURI();
	}
}
```

**Comment by pieterjanpintens** [comment]
> Can you try with 9.4.24? I recall we fixed a bug in this area recently. I'm away from my computer right now, but will post a link to the issue tomorrow.

I will try that later, we rather stay on the version that comes with the eclipse release but if this release solves this problem we can diverge.

**Comment by janbartel** [comment]
@pieterjanpintens  this was the bug I was thinking of, which sounds extremely relevant to your situation: https://github.com/eclipse/jetty.project/issues/4183

**Comment by pieterjanpintens** [comment]
Yes we think this will resolve our problem! Thank you
